src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,508 @@
+import { CssClassesOnly, Transition } from './transitions';
+import { DefaultMap } from '../../map/default-map';
+import { mathf, Raf } from '../..';
+import { setf } from '../../setf/setf';
+import { CarouselSynchronizer } from './carousel-synchronizer';
+
+/**
+ * For readability, specifies which half should be gotten by providing a
+ * direction to loop through elements.
+ */
+enum Half {
+  LEFT = -1,
+  RIGHT = 1
+}
+
+/**
+ * Options provided to the constructor
+ */
+export interface CarouselOptions {
+  onTransitionCallbacks?: Array<(carousel: Carousel) => void>;
+  onDisposeCallbacks?: Array<(carousel: Carousel) => void>;
+  activeCssClass?: string;
+  beforeCssClass?: string;
+  afterCssClass?: string;
+  allowLooping?: boolean;
+  // Transition used to iterate through slides.
+  transition?: Transition;
+  // Carousel code won't run if the condition is not met.
+  // Used for establishing multiple types of carousels on the same DOM varied
+  // by breakpoint.
+  condition?: () => boolean;
+}
+
+/**
+ * Default classes applied to slide elements.
+ */
+enum DefaultCssClass {
+  ACTIVE_SLIDE = 'active',
+  BEFORE_SLIDE = 'before',
+  AFTER_SLIDE = 'after'
+}
+
+/**
+ * Tracks an element that should be transitioned to.
+ */
+class TransitionTarget {
+  readonly element: HTMLElement;
+  readonly drivenBySync: boolean;
+
+  constructor(element: HTMLElement, drivenBySync: boolean = false) {
+    this.element = element;
+    this.drivenBySync = drivenBySync;
+  }
+}
+
+export class Carousel {
```

Comment:
More of an overall comment on the Carousel structure, I usually prefer a simple structure for carousels and UI components (as you may have seen on some of our projects) - usually a single class without much cross dependency.

There seems to be a lot of manual calculations (linear alg, vector, matrices) which I think when we are doing things fairly complicated like a super custom hero or webGL stuff but where possible for something like a carousel, I think it might be best to just keep it simple and learn on the browser to do it's work.

However, this might be because your carousel has provides advantages that I don't quite see yet.  Can you help me understand how `adjust-slide-for-loop`, `adjust-slide-for-split`, `physical-slide` all kind of fit together here?

Reply:
Of course. I prefer simpler too and I'll see what else I can streamline or incorporate tighter with existing Degu code.

That said, this started much simpler and more akin to what we've seen in other projects. First lines were committed back in March of 2018, but the complexity has been built up to address corner cases and performance issues encountered across projects.

This carousel for instance, needed transitions based on slide position updated in real time. That's where MatrixService played in, updating the scale of the device with transform information that would be changed by the time the scale change was rendered resulted in a jitter. This necessitated referencing translation changes made in the read step, to perform other calculations before making adjustments during the write step. https://store.google.com/us/product/google_nest_mini#mini-solutions-audio

`adjustSlideForLoop` fixes the need for duplicating slides in an infinite loop situation, allowing for slides to transform independently outside of their container. I know the example didn't reflect this, I found the bug during the port and its been rectified. In short, if a slide has moved far enough offscreen to either side it puts it on the opposite side. I'll see if I can find a better name and better document it in code.

`adjustSlideForSplit` Ensures that other slides keep the appropriate distance from the active slide.

`physicalSlide` handles user interaction for dragging slides. The name should probably be updated now that I'm thinking about it. At one point, probably years ago now, it had friction/velocity/etc. and was "physical" that way, which it isn't anymore.

I'll take another polish pass through and see if I can better explain/justify what exists and see if there's anything further I can remove without hurting performance.